website/docs/type-annotations.md

```diff
@@ -86,29 +86,42 @@ def current_user
 end
 ```
 
-## Why do I need to repeat types from the constructor?
+## Why can't Sorbet infer the types my instance variables?
 
-A current shortcoming of Sorbet is it cannot reuse static type knowledge in
-order to automatically determine the type of an instance or class variable. In
-the following example, despite the fact that Sorbet knows that `x` has type
-`Integer`, it still treats `@x` as `T.untyped` without an explicit type
-annotation:
+A current shortcoming of Sorbet is in many cases it cannot reuse static type
+knowledge in order to automatically determine the type of an instance or class
+variable. In the following example, Sorbet will naturally understand that `@x`
+is of type `Integer`, but it cannot determine the static type of `@y` without a
+`T.let` and therefore treats it as `T.untyped`:
```

Comment:
This section comes up somewhat unexpectedly.

There's already a line above that's like

> The exception to the above list is that Sorbet does not need type annotations for local variables.

What do you think about adding a new section near the top of the document titled "When are type annotations optional?" or "When are type annotations required?" and maybe forward reference to this section which has more complete examples?

Reply:
I've added this section and talked about this, and also lightly touched on when you might want to use a typed local for loop reasons, which seems like it's related: https://github.com/sorbet/sorbet/pull/2230/files#diff-869efd8b0cd4a253689d674e63e6d6bcR24-R65